rfc/rfc-0000-allow-resources-to-report-status.md

```diff
@@ -0,0 +1,320 @@
+For `matchFields`, the spec is based off of `matchFields` in blueprint resource options. Authors need to define the `key` and `operator`, and potentially the `values` (if the operator is `In` or `NotIn`). Authors can also optionally define a `messagePath`, which is a path in the resource where Cartographer can pull off a message to display in the owner status.
+
+Each template will now have a new field in the spec `resourceStatusRuleRef` where authors can specify the `ClusterResourceStatusRule` to use for that template. If no `ClusterResourceStatusRule` is defined, Cartographer will default to listing the resource as `Succeeded` once it has been successfully applied to the cluster and any relevant outputs have been read off the resource.
+
+Each owner will display a new condition in the status called `ResourcesSuceeded` which will only be updated to `True` when all resources in the blueprint have succeeded. Cartographer will not update the top level state of an owner to `Ready: True` until `ResourcesSucceeded` is `True`.
+
+# Migration
+[migration]: #migration
+
+Given that Cartographer will default to `Succeeded` once the resource is applied and outputs are read, this will not be a breaking change.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+- The owner status will potentially grow significantly, based on how many matchers are in a `ClusterResourceStatusRule`.
+- We only allow for reflecting three states, though this design allows for creating more states in the future.
+- This design does not allow authors to define the states themselves.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- What other designs have been considered?
+  - The `ClusterResourceStatusRule` could instead be inlined in the templates.
+    - Separating these rules in a separate CRD allows for re-usability of those rules.
+  - We could also not require the templates to specify the `ClusterResourceStatusRule` and instead have those rules automatically select templates based on the `kind` and `apiVersion` of the resource they are stamping out.
+    - This could be a bit confusing because of ytt. A template could end up stamping out different types of resources and so Cartographer would only be able to inform users which `ClusterResourceStatusRule` is used for a given template after an owner is applied, and it could then be reflected in the owner status.
+- What is the impact of not doing this?
+  - Users/UI creators will have to have in depth knowledge of the underlying resources to determine the health of the owner.
+
+# Prior Art
+[prior-art]: #prior-art
+
+N/A
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- Should the two states be `Succeeded` and `Failed`?
+- Is it right for Cartographer to default to `Succeeded`?
```

Comment:
`Succeed` by default indeed makes sense for some resources like `configmap`, but I'm not very sure that's a good default overall - I feel like we should default to `Unknown` as per the description of it

> _"Unknown" strictly represents that Cartographer has not be able to determine success or failure._

it sounds quite fit for that 🤔  

There'd definitely be the implication that the next line talks about (`Ready: True` only if `ResourcesSucceeded: True`) if we go with defaulting to Unknown though (it'd make this a breaking change).

---

any thoughts from folks around defaulting to Unknown instead?

Reply:
We have moved to using `healthy` true/false/unknown, but this comment is still relevant. Interested to hear what others think about this.

Reply:
Lets move this conversation to the one started below by @scothis